src/uct/ib/mlx5/ib_mlx5.c

```diff
@@ -543,48 +543,109 @@ int uct_ib_mlx5_devx_uar_cmp(uct_ib_mlx5_devx_uar_t *uar,
+    const char *allocation_type            = (flags == UCT_IB_MLX5_UAR_ALLOC_TYPE_WC) ? "WC" : "NC_DEDICATED";
+    ucs_log_level_t err_log_level          = UCS_LOG_LEVEL_DIAG;
     struct mlx5dv_devx_uar *uar;
-    char buf[512];
+    ucs_status_t status;
+    ucs_string_buffer_t buf;
+    int err;
 
     uar = mlx5dv_devx_alloc_uar(md->super.dev.ibv_context, flags);
     if (uar == NULL) {
-        sprintf(buf, "mlx5dv_devx_alloc_uar(device=%s, flags=0x%x(%s)) "
-                "failed: %m", uct_ib_device_name(&md->super.dev), flags, title);
-        if (fallback == NULL) {
-            ucs_log(log_level, "%s", buf);
-        } else {
-            ucs_log(log_level, "%s, fallback to %s", buf, fallback);
+        ucs_string_buffer_init(&buf);
+
+        err = errno;
+        ucs_string_buffer_appendf(&buf, "mlx5dv_devx_alloc_uar(device=%s, flags=0x%x) type=%s failed: %s. ",
+                                  uct_ib_device_name(&md->super.dev), flags, allocation_type, strerror(err));
+        
+        switch (err) {
+            case ENOMEM:
+                /* UAR allocation failed due to insufficient PF_LOG_BAR_SIZE; fallback will produce the same result */
+                ucs_string_buffer_appendf(&buf, "%s", pf_log_bar_size_suggestion);
+                err_log_level = UCS_LOG_LEVEL_ERROR;
+                status        = UCS_ERR_NO_MEMORY;
+                break;
+
+            case EOPNOTSUPP:
+                /* UAR allocation failed due to unsupported allocation type, if there's no fallback, it's an error */
+                status = UCS_ERR_UNSUPPORTED;
+                break;
+
+            case EINVAL:
+                status = UCS_ERR_INVALID_PARAM;
+                break;
+
+            default:
+                status = UCS_ERR_NO_MEMORY;
+                break;
         }
-
-        return UCS_ERR_NO_MEMORY;
+        ucs_log(err_log_level, "%s", ucs_string_buffer_cstr(&buf));
+        ucs_string_buffer_cleanup(&buf);
+        return status;
     }
 
     *uar_p = uar;
     return UCS_OK;
 }
 #endif
 
+ucs_status_t uct_ib_mlx5_devx_check_supported_uar_alloc_type(uct_ib_mlx5_md_t *md)
+{
+#if HAVE_DEVX
+    uct_ib_mlx5_devx_uar_t uar;
+    ucs_status_t status;
+
+    status = uct_ib_mlx5_devx_alloc_uar(md, UCT_IB_MLX5_UAR_ALLOC_TYPE_WC, &uar.uar);
+    if (status == UCS_ERR_UNSUPPORTED) {
+        ucs_debug("WC UAR allocation type is not supported, falling back to NC_DEDICATED");
+        
+        status = uct_ib_mlx5_devx_alloc_uar(md, UCT_IB_MLX5_UAR_ALLOC_TYPE_NC_DEDICATED, &uar.uar);    
+        if (status == UCS_ERR_UNSUPPORTED) {
+            ucs_error("Both WC and NC_DEDICATED UAR allocation types are not supported");
+            return status;
+        } else if (status != UCS_OK) {
+            return status;
+        } 
+
+        /* NC_DEDICATED is supported - set the flag to 0 */
+        md->flags &= ~UCT_IB_MLX5_MD_FLAG_UAR_ALLOC_TYPE;
```

Comment:
no need, it is 0 before calling this function

Reply:
done